File: 2023/day2.R

# perfect ticyverse solution from @drob
# https://twitter.com/drob/status/1730817227371225487

library(tidyverse)

# have to load d first

d = read_lines("2023/input2.txt") |>
    as_tibble() |>
    rename(x = value)

game_cubes = d |>
    mutate(game = row_number()) |>
    mutate(cubes = str_extract_all(x, "\\d+ [a-z]+")) |>
    unnest(cubes) |>
    separate(cubes, c("number", "color"), sep = " ", convert = TRUE)


# part 1

maxima = c(red = 12, green = 13, blue = 14)

game_cubes |>
    mutate(maximum = maxima[color]) |>
    group_by(game) |>
    summarise(playable = !any(number > maximum)) |>
    filter(playable) |>
    summarise(sum(game))


# part 2

game_cubes |>
    group_by(game, color) |>
    summarise(number = max(number)) |>
    summarise(power = prod(number)) |>
    summarise(sum(power))

File: 2023/day2.py
from collections import defaultdict
from functools import reduce
from itertools import product

# part 1

result = 0
with open("input2.txt") as f:
    lines = f.readlines()
    for line in lines:
        line = line.strip()
        game, bag = line.split(": ")
        number = game.split()[1]
        sts = bag.split("; ")
        valid = True
        for st in sts:
            counts = defaultdict(int)
            for cubes in st.split(", "):
                count, color = cubes.split()
                counts[color] += int(count)
            if counts["red"] > 12 or counts["green"] > 13 or counts["blue"] > 14:
                valid = False
        if valid:
            result += int(number)

    print(result)



# part 2

result = 0
with open("input2.txt") as f:
    lines = f.readlines()
    for line in lines:
        line = line.strip()
        game, bag = line.split(": ")
        number = game.split()[1]
        sts = bag.split("; ")

        counts = defaultdict(int)
        for st in sts:
            for cubes in st.split(", "):
                count, color = cubes.split()
                counts[color] = max(int(count), counts[color])
        result += reduce(lambda x, y: x * y, counts.values())


    print(result)






